siliconcompiler/schema/schema_obj.py

```diff
@@ -194,204 +229,162 @@ def record_history(self):
-                        sctype = re.sub(r'[\[\]]', '', cfg[param]['type'])
-                        return_list = []
-                        if selval is None:
-                            return None
-                        for item in selval:
-                            if sctype == 'int':
-                                return_list.append(int(item))
-                            elif sctype == 'float':
-                                return_list.append(float(item))
-                            elif sctype.startswith('(str,'):
-                                if isinstance(item,tuple):
-                                    return_list.append(item)
-                                else:
-                                    tuplestr = re.sub(r'[\(\)\'\s]','',item)
-                                    return_list.append(tuple(tuplestr.split(',')))
-                            elif sctype.startswith('(float,'):
-                                if isinstance(item,tuple):
-                                    return_list.append(item)
-                                else:
-                                    tuplestr = re.sub(r'[\(\)\s]','',item)
-                                    return_list.append(tuple(map(float, tuplestr.split(','))))
-                            else:
-                                return_list.append(item)
-                        return return_list
-                    else:
-                        if selval is None:
-                            # Unset scalar of any type
-                            scalar = None
-                        elif cfg[param]['type'] == "int":
-                            #print(selval, type(selval))
-                            scalar = int(float(selval))
-                        elif cfg[param]['type'] == "float":
-                            scalar = float(selval)
-                        elif cfg[param]['type'] == "bool":
-                            scalar = (selval == 'true')
-                        elif re.match(r'\(float', cfg[param]['type']):
-                            tuplestr = re.sub(r'[\(\)\s]','',selval)
-                            scalar = tuple(map(float, tuplestr.split(',')))
-                        elif re.match(r'\(str', cfg[param]['type']):
-                            tuplestr = re.sub(r'[\(\)\'\s]','',selval)
-                            scalar = tuple(tuplestr.split(','))
-                        else:
-                            scalar = selval
-                        return scalar
-                #all non-value fields are strings (or lists of strings)
+            if Schema._is_leaf(cfg):
+                raise ValueError(f'Invalid keypath {keypath}: unexpected key: {key}')
+
+            if key in cfg:
+                cfg = cfg[key]
+            elif 'default' in cfg:
+                if insert_defaults:
+                    cfg[key] = copy.deepcopy(cfg['default'])
+                    cfg = cfg[key]
```

Comment:
A beautiful example of replacing technical debt (mine) with a much cleaner and better implementation. Thank you!!